sdk/src/Core/Amazon.Runtime/Credentials/FallbackCredentialsFactory.cs

```diff
@@ -124,61 +136,84 @@ public static AWSCredentials GetCredentials()
+                    }
+                    // Breaking the FallbackCredentialFactory chain in case a ProcessAWSCredentialException exception 
+                    // is encountered. ProcessAWSCredentialException is thrown by the ProcessAWSCredential provider
+                    // when an exception is encountered when running a user provided process to obtain Basic/Session 
+                    // credentials. The motivation behind this is that, if the user has provided a process to be run
+                    // he expects to use the credentials obtained by running the process. Therefore the exception is
+                    // surfaced to the user.
+                    catch (ProcessAWSCredentialException)
+                    {
+                        throw;
+                    }
+                    catch (Exception e)
+                    {
+                        cachedCredentials = null;
 
-                if (cachedCredentials != null)
-                    break;
-            }
+                        errors.Add(e);
+                    }
 
-            if (cachedCredentials == null)
-            {
-                if (fallbackToAnonymous)
-                {
-                    return new AnonymousAWSCredentials();
+                    if (cachedCredentials != null)
+                        break;
                 }
 
-                using (StringWriter writer = new StringWriter(CultureInfo.InvariantCulture))
+                if (cachedCredentials == null)
                 {
-                    writer.WriteLine("Unable to find credentials");
-                    writer.WriteLine();
-                    for (int i = 0; i < errors.Count; i++)
+                    if (fallbackToAnonymous)
                     {
-                        Exception e = errors[i];
-                        writer.WriteLine("Exception {0} of {1}:", i + 1, errors.Count);
-                        writer.WriteLine(e.ToString());
-                        writer.WriteLine();
+                        return new AnonymousAWSCredentials();
                     }
 
-                    throw new AmazonServiceException(writer.ToString());
+                    using (StringWriter writer = new StringWriter(CultureInfo.InvariantCulture))
+                    {
+                        writer.WriteLine("Unable to find credentials");
+                        writer.WriteLine();
+                        for (int i = 0; i < errors.Count; i++)
+                        {
+                            Exception e = errors[i];
+                            writer.WriteLine("Exception {0} of {1}:", i + 1, errors.Count);
+                            writer.WriteLine(e.ToString());
+                            writer.WriteLine();
+                        }
+
+                        throw new AmazonServiceException(writer.ToString());
+                    }
                 }
-            }
 
-            return cachedCredentials;
+                return cachedCredentials;
```

Comment:
we should exit the lock first then return cached credentials

Reply:
The finally block will ensure that. Otherwise we can return a null value here as it can be overwritten.

Reply:
Ah I misread the scope here, I thought we were returning outside the try block